### PR TITLE
docs(roadmap): define Application composition and State

### PR DESCRIPTION
**Pull request by paulfalgout**
## Linked issue

- Related #190
- Related #191

## What

- Strengthen `Application` as the single root-or-nested lifecycle and ownership scope without adding a parallel Feature class.
- Define `State` as a first-class pay-for-play object composed into Marionette owners rather than inherited or mixed into them.
- Preserve `MnObject` as an optional passive convenience while preferring plain classes when Marionette lifetime conventions are unnecessary.
- Document Toolkit App/Component migration decisions, View-only Region ownership, deterministic Application/Region teardown, and the reference-app overlay-host pattern.

## Agent-development failure addressed

The previous roadmap covered rendered View/Region ownership but did not provide nested Application structure or an ergonomic local-state contract. An intermediate proposal added `Feature`, but Application and Feature had nearly identical responsibilities and created unnecessary naming, migration, and inheritance complexity. This revision keeps Marionette's existing Application vocabulary and makes composition—not class hierarchy—the canonical architecture.

## Public behavior contract

- Canonical behavior after this PR: Phase 1 will strengthen `Application` for root and nested scopes, compose first-class `State` where requested, keep `MnObject` optional and passive, keep Region operations View-only, and specify deterministic lifecycle and teardown behavior before implementation.
- Superseded behavior removed, if any: the intermediate roadmap proposal for a separate `Feature` class is removed.
- Compatibility exception and removal condition, if any: none; this PR changes the release plan rather than runtime behavior.

## Runtime cost

- [x] Static or documentation only
- [ ] Development or test only; excluded from production entrypoints
- [ ] Changes an existing production path
- [ ] Adds an explicitly opt-in runtime path

Measured impact or explanation:

- Bundle: no runtime source or production artifact changes; hosted report shows 0 B change.
- Startup/hot path: no runtime implementation in this PR.
- Allocations/retention: no runtime implementation in this PR; the roadmap requires zero State or child storage until first use.

## Validation

- [ ] Tests cover the acceptance criteria and relevant edge cases.
- [x] Docs, examples, declarations, diagnostics, and rule metadata agree where relevant.
- [x] No private consumer, private fixture, or undocumented context is required.
- [x] I ran the commands listed below and am reporting their actual results.

Commands:

```sh
git diff --check
awk 'length($0) > 100 { print FNR ":" length($0) ":" $0 }' ROADMAP.md
```

Both commands completed without findings. Runtime, browser, and package tests were not run locally because this PR changes only the governing Markdown strategy. GitHub CI validates the repository at the exact head.

## Package and browser impact

- Entry points or install fixtures affected: none.
- Browser contracts affected: none in this PR; the roadmap requires later specification and validation.
- Production/dev/test module-graph impact: none.

## Rollback or deprecation

Revert the roadmap commits. No runtime API, persisted data, package, deployment, or public website content changes.

## Review notes

- The branch name retains the discarded `feature` wording, but the public roadmap, PR title, and tracker use `Application`; renaming the branch would add ref churn without changing the proposal.
- Three bounded Claude architecture reviews clarified independent first-class classes, State composition, View-only Region operations, one-way Application/Region teardown, root/nested API parity, Behavior state ownership, and Toolkit Component migration decisions.
- Exact-head Claude and Cubic feedback corrected stale Feature/local-state vocabulary, Behavior State persistence, asynchronous start settlement requirements, and benchmark capability coverage.
- This PR does not publish documentation to the public website.

## Summary by CodeRabbit

* **Documentation**
  * Updated the roadmap to define Applications as the primary lifecycle and ownership scope.
  * Clarified how nested Applications, Views, Regions, and optional State objects work together.
  * Added guidance for state allocation, toolkit migration, and overlay hosting.
  * Expanded planning and benchmark requirements to cover Applications, State, ownership, and overlays.

### CODE REVIEW

**Comment by paulfalgout** [comment]
Reviewed at `15233a69` (the SHA I fetched). ROADMAP-only, +93/−10. The architecture is carefully drawn — in particular the Feature/Region boundary holds the line that was easy to lose: *"Region remains the only object that mounts or tears down a hosted View"*, with Feature stop allowed to ask its host Region to empty only while that Region still holds the Feature's root View, and external replacement clearing the stale reference without implicitly stopping an independently active Feature. That's the dual-ownership failure mode named and closed rather than left to Phase 1. Same for the pay-for-play state contract, which is stated in allocation terms ("no store, listener, cleanup registration, or instance property until state is declared, supplied, or first mutated") and backed by a matching Phase 0 allocation test. **Approve** — two things I'd want decided deliberately rather than absorbed.

**1. The corpus grew; the task floor didn't.** The required coverage is now plain Views, stateful Views, Regions, Features, lifecycle repair, Behaviors, shared-host overlays, communication boundaries, and cleanup — call it nine capability areas. The floor is still *"at least 10 tasks."* At that floor it's roughly one task per area with no redundancy, and that interacts badly with two gates already in this section:

- *"no individual task has a fully-correct point estimate below 60 percent"* becomes a per-**area** gate by accident. One awkwardly-worded prompt or unlucky evaluator in the overlay task blocks release and reads as "overlays are weak," with no second task in that area to distinguish task noise from a real capability gap. The converse also holds: an easy task silently certifies its whole area.
- Sample-size planning uses *"0.05 divided by the task count."* Raising the count to cover the new areas tightens per-task alpha, which raises the paired-run count needed for the same 80 percent power. So broadening coverage costs runs superlinearly, and the doc expands the breadth requirement without revisiting either the floor or that cost.

Worth stating a floor that scales with the areas — two tasks per area, or an explicit "areas may share tasks" — so the per-task floor isn't silently doing per-area duty.

**2. "Singular" is asserted in one section and qualified in another.** Under *Coherent runtime contracts*: *"`Application` is the singular, non-nestable root Feature."* The qualifier lands much later, under *Architecture boundaries*: *"Root singularity applies per composition tree; independent Application instances may coexist on a page."* Those reconcile, but a reader consulting the contracts list — the more likely destination for "can I have two Applications?" — gets only the stronger claim. Since the stated goal is that agents read contracts without inferring, I'd put "per composition tree" at the first assertion.

Neither blocks. I did check the two ROADMAP inconsistencies I'd raised earlier in this cycle, and both are gone on this branch: Brotli-11 is now used consistently with no lingering minified-gzip reference, and the extension-hook lines at 204/322/344 now read coherently across the phases.

<sub>Reviewed with Claude Code.</sub>

**Comment by paulfalgout** [comment]
Re-reviewed at `d77a3dd0` (the SHA I fetched). **Approve.** Deleting `Feature` outright is a better answer than the one I was reviewing — the previous head had `MnObject → Feature → Application` with Application defined as a non-nestable subtype, which needed a special case for the root. Identifying the root by parentlessness removes the class *and* the special case, and "Root status is topology, not a reason for an Application subtype" is the sentence that keeps it from growing back.

This also resolves my second point from the last head: the unqualified *"singular, non-nestable root Feature"* is gone, and the qualifier now sits with the claim rather than four sections away.

The new teardown ordering is the strongest part. Making stop *ordered* — children, then host Region only if it still holds the root View, then resources — plus *"a later stop is idempotent and cannot empty an unrelated replacement View"* closes the race where a Region is re-shown between an external `empty` and a subsequent stop. That's the failure that would otherwise be found in an application rather than in a test.

**Two things this refactor left behind.**

**1. Line 118 still requires documenting a class that no longer exists.**

```
- A compact agent-oriented reference describes lifecycle, ownership, regions,
  features, local state, behaviors, communication, and teardown without inventing a
  separate API.
```

`features` was added in the previous head to name the Feature contract, and `local state` is the pre-rename term for `State`. As written, the Phase 2 deliverable instructs whoever writes that reference — plausibly an agent, given the stated audience — to document a Feature concept the architecture section now explicitly forbids. Reads as `applications, State`. Line 220's *"the local-state or data-source contracts"* has the same stale proper noun. The Feature mentions at 225 and 332 are deliberate negations and correctly left alone.

**2. Behavior became a State owner without a persistence rule.** Behavior is now in the eligible set and in the Phase 0 allocation tests, but the persistence enumeration covers only three of the four:

> State composed into View or CollectionView persists across render; State composed into Application persists across stop and restart; State composed into MnObject persists for the object's lifetime; every owned State ends at owner destroy.

Behavior isn't there. It isn't inferable either — a Behavior outlives its View's re-render, so "persists across render" is a plausible rule, but so is treating it as MnObject-like for its attachment lifetime, and the two differ if a Behavior is ever re-attached. Since this doc's purpose is that agents read contracts instead of inferring them, an eligible owner with no stated rule is the gap most likely to produce two reasonable, incompatible implementations.

**Still open from last time:** the corpus breadth versus the unchanged "at least 10 tasks" floor. The rename to Applications doesn't change the arithmetic. Noting it once more only because the paragraph was otherwise edited in this push; I won't raise it again.

<sub>Reviewed with Claude Code.</sub>

File: ROADMAP.md [hunk 1]
```diff
@@ -78,8 +82,17 @@ budget.
 ### Coherent runtime contracts
 
 - Lifecycle methods and events have documented state machines and ordering.
+- `MnObject` is the minimal non-renderable, evented, destroyable object; `Feature`
+  adds an active lifecycle and owned composition; `Application` is the singular,
+  non-nestable root Feature.
+- Feature parent/child ownership, root View and Region association, asynchronous
+  startup, restart invalidation, and teardown are explicit and testable.
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply ROADMAP.md]: Handled in bfd557c: the roadmap and #190 now require an explicit start return/readiness/failure contract, deterministic overlap settlement, and migration coverage before implementation.

File: ROADMAP.md [hunk 2]
```diff
@@ -235,6 +308,13 @@ instructions, and every release blocker maps to this strategy.
 ### Phase 1: Core contracts
 
 - Specify lifecycle state machines and add transition-table or model-based tests.
+- Introduce Feature as the nested non-renderable lifecycle and ownership scope while
+  keeping Application the singular, non-nestable root Feature and MnObject the
+  minimal base. Specify deterministic asynchronous startup, stale-start invalidation,
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply ROADMAP.md]: Handled in bfd557c: the roadmap and #190 now require an explicit start return/readiness/failure contract, deterministic overlap settlement, and migration coverage before implementation.

**Comment by paulfalgout** [comment]
Re-reviewed at `bfd557c3` (the SHA I fetched). **Approve.** All three open points are closed, and I checked rather than assuming: `grep -nE "local.state|features,"` over the file now returns nothing, so the rename is complete rather than partial.

- Line 118 reads `Applications, State`, and lines 220 and 351 use `State` as the proper noun.
- Behavior got its persistence rule — *"persists across its owning View's render and ends when the Behavior is destroyed"* — which is the resolution I'd have picked, since a Behavior outliving re-render but not its own destroy is the only reading consistent with how Behaviors are torn down.
- *"Every named capability area is exercised by at least two independently scored tasks. One task may cover multiple areas, but no single task certifies an area."* This is a better fix than the floor increase I suggested: it constrains the property that actually mattered (no area resting on one task) instead of a proxy for it, and the multi-area allowance keeps it satisfiable near the existing floor — roughly ten areas at two tasks each is reachable with ~10 tasks averaging two areas apiece, so the Bonferroni cost of more tasks isn't forced.

Two additions in this push worth noting on their own:

**Region ownership.** *"An Application destroys a Region it owns, but never destroys a borrowed host Region."* That's the natural companion to the existing empty-only-if-it-still-holds-my-View rule — together they cover both ways an Application could reach past its boundary into a Region it didn't create.

**The `start` contract.** Requiring Phase 1 to pin `start`'s return value, readiness signal, failure propagation, and how overlapping start/stop/restart settle is the right thing to force early. The clause I'd single out is *"migration tests must cover the existing synchronous return contract"* — `start` returning synchronously today means any move to a promise-shaped result is a breaking change for existing callers, and naming it here keeps that from surfacing as a surprise mid-Phase-1. *"Settle deterministically without exposing stale success or leaving callers pending"* also rules out the failure mode where an invalidated start neither resolves nor rejects.

Nothing further from me.

<sub>Reviewed with Claude Code.</sub>

File: ROADMAP.md [hunk 3]
```diff
@@ -145,17 +166,101 @@ budget.
...
+`State` is a first-class, event-notifying object composed into an eligible owner rather
+than methods mixed into several unrelated classes. Phase 1 selects its concrete API
+after testing the current Toolkit vocabulary against representative plain and
+stateful objects. No State instance, listener, cleanup registration, or owner
+property exists until state is declared, supplied, or first requested. State composed
+into View or CollectionView persists across render; State composed into Application
+persists across stop and restart; State composed into MnObject persists for the
+object's lifetime; State composed into Behavior persists across its owning View's
+render and ends when the Behavior is destroyed. Every owned State ends at owner
+destroy. An invalidated asynchronous start cannot later mutate State or owned
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply ROADMAP.md]: Handled in bd110e1: supplied unowned State transfers to one owner, already-owned State is rejected, shared data uses an external data source, and #191 requires transfer/rejection/cleanup tests.

**Comment by paulfalgout** [comment]
Re-reviewed at `bd110e10` (the SHA I fetched). **Approve.** Single-ownership is the right rule and closes a hazard the previous head left open: without it, two owners could each hold the same State and each destroy it at their own teardown, so lifetime depended on which one died first. *"Supplying an unowned State transfers ownership; supplying an already-owned State fails with a stable diagnostic"* makes that a startup error instead of a teardown mystery, which is the correct place to fail.

**One contradiction this introduced.** The paragraph's closing sentence wasn't reconciled with the new rule:

> A Behavior may compose private State when its concern truly owns that state; state shared with its View belongs on the View and **is passed to the Behavior**.

If the View owns that State, passing it to the Behavior is supplying an already-owned State — which the same paragraph now says fails with a stable diagnostic. The two readings that would rescue it are both unstated: either "passed" means the Behavior holds a plain reference and reads/writes through it without composing it as its own State, or there's a borrow concept that *"rather than sharing or borrowing it"* explicitly rules out. Since this is the sentence someone consults for exactly this case, it needs to say which. Something like *"…belongs on the View, and the Behavior receives a reference to it without taking ownership"* would resolve it, if that's the intent.

**Trivial, same paragraph.** The reflow left two artifacts: the line break after *"State composed into Behavior persists across its owning View's"* wraps mid-phrase and leaves a short line, and the persistence list now ends its first item with a period while the rest stay semicolon-joined, so what reads as one list is punctuated as two. Worth a pass whenever the sentence above is edited.

<sub>Reviewed with Claude Code.</sub>

**Comment by paulfalgout** [comment]
Re-reviewed at `ca0eb824` (the SHA I fetched). **Approve — nothing open from me.**

The contradiction is resolved from both ends rather than just the one I pointed at, which is the better fix. The rule now scopes itself — *"supplying an already-owned State **for composition** fails"* — so it says what it governs instead of leaving the reader to infer it, and the closing sentence states the permitted case explicitly: *"the Behavior may receive an ordinary reference to that State but does not compose, own, or destroy it."* Either edit alone would have left the other sentence carrying an unstated qualifier.

*"rather than creating shared or borrowed ownership"* is also a real improvement over the previous *"rather than sharing or borrowing it"* — what can't be shared is ownership, not the reference, and the old phrasing read as forbidding the very thing the last sentence now permits.

Reflow artifacts are gone too: all four persistence clauses are period-separated consistently, and the mid-phrase break is fixed.

<sub>Reviewed with Claude Code.</sub>
